vnext/Microsoft.ReactNative/JsiWriter.cpp

```diff
@@ -0,0 +1,146 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "pch.h"
+#include "JsiWriter.h"
+#include <crash/verifyElseCrash.h>
+
+namespace winrt::Microsoft::ReactNative {
+
+//===========================================================================
+// JsiWriter implementation
+//===========================================================================
+
+JsiWriter::JsiWriter(facebook::jsi::Runtime &runtime) noexcept : m_runtime(runtime) {
+  Push({ContinuationAction::AcceptValueAndFinish});
+}
+
+facebook::jsi::Value JsiWriter::MoveResult() noexcept {
+  VerifyElseCrash(m_continuations.size() == 0);
+  return std::move(m_result);
+}
+
+facebook::jsi::Value JsiWriter::CopyResult() noexcept {
+  VerifyElseCrash(m_continuations.size() == 0);
+  return {m_runtime, m_result};
+}
+
+void JsiWriter::WriteNull() noexcept {
+  WriteValue(facebook::jsi::Value::null());
+}
+
+void JsiWriter::WriteBoolean(bool value) noexcept {
+  WriteValue({value});
+}
+
+void JsiWriter::WriteInt64(int64_t value) noexcept {
+  // JavaScript's integer is not int64_t, need to ensure that the value is in range
+  VerifyElseCrash(MinSafeInteger <= value && value <= MaxSafeInteger);
+
+  // unfortunately JSI only supports int and double for number, choose double here
+  // make sure that doing type conversion in C++ makes no data loss
+  double d = static_cast<double>(value);
+  VerifyElseCrash(floor(d) == d);
```

Comment:
>VerifyElseCrash [](start = 2, length = 15)

Do not forget that some of the data maybe coming from user. We should not crash on lost precision.